forest/bonsai/simulate_gps_data.py

```diff
@@ -1117,31 +1123,306 @@ def process_switches(
+            format "Country_2_letter_ISO_code/City_Name"
+        start_date: (datetime.date) start date of trajectories
+        end_date: (datetime.date) end date of trajectories,
+            end date is not included in the trajectories
+        cycle: (int) the sum of on-cycle and off_cycle,
+            unit is minute
+        percentage: (float) the missing rate, in other words,
+            the proportion of off_cycle, should be within [0,1]
+        api_key: (str), api key for open route service
+            https://openrouteservice.org/
+        attributes_dict: (dictionary) containing attributes
+            for each user, optional
+    Returns:
+        gps_data: (pandas.DataFrame) contains gps trajectories
+            for each person
+    Raises:
+        ValueError: if attributes.json is not in the correct format
+        ValueError: if location is not in the correct format
+        RuntimeError: if too many Overpass queries are made
+        ValueError: if Overpass query fails
+    """
+
+    sys.stdout.write("Loading Attributes...\n")
+
+    if attributes_dict is None:
+        attributes_dictionary: Dict[int, Attributes] = {}
+        switches_dictionary: Dict[int, Dict[str, int]] = {}
+    else:
+        attributes_dictionary, switches_dictionary = load_attributes(
+            attributes_dict
+        )
+
+    for user in range(1, n_persons + 1):
+        if user not in attributes_dictionary.keys():
+            attributes_dictionary[user] = Attributes()
+        if user not in switches_dictionary.keys():
+            switches_dictionary[user] = {}
+
+    sys.stdout.write("Gathering Addresses...\n")
+    try:
+        location_ctr, location_city = location.split("/")
+    except ValueError:
+        raise ValueError("Location provided did not have the correct format.")
+
+    nodes = generate_addresses(location_ctr, location_city)
+
+    # find timezone of city
+    location_coords = (float(nodes[0]['lat']), float(nodes[0]['lon']))
+
+    obj = TimezoneFinder()
+    tz_str = obj.timezone_at(lng=location_coords[1], lat=location_coords[0])
+
+    no_of_days = (end_date - start_date).days
+
+    timestamp_s = (
+        datetime2stamp(
+            [start_date.year, start_date.month, start_date.day, 0, 0, 0],
+            tz_str
+        )
+        * 1000
+    )
+
+    user = 0
+    ind = 0
+    gps_data = pd.DataFrame(columns=[
+        "user",
+        "timestamp",
+        "UTC time",
+        "latitude",
+        "longitude",
+        "altitude",
+        "accuracy",
+        ]
+    )
+    sys.stdout.write("Starting to generate trajectories...\n")
+    while user < n_persons:
+
+        house_address = (float(nodes[ind]['lat']), float(nodes[ind]['lon']))
+        attrs = attributes_dictionary[user + 1]
+
+        all_nodes = generate_nodes(house_address, attrs.main_occupation)
+
+        person = Person(house_address, attrs, all_nodes)
+        all_traj, all_T, all_D = gen_all_traj(
```

Comment:
Variable names should be lowercase and it'd great to expand the `T` and `D` to help with readability.